django-rgd/rgd/rest/viewsets.py

```diff
@@ -18,6 +19,24 @@ class CollectionViewSet(ModelViewSet):
     queryset = models.Collection.objects.all()
     filterset_class = CollectionFilter
 
+    @swagger_auto_schema(
+        method='GET',
+        operation_summary='Get the file at the given index in this collection. Associated files are orderd by primary key and the provided index is the index in that ordered set. If the files in the collection change, this will not be reproducible.',
+    )
+    @action(
+        detail=True,
+        serializer_class=serializers.ChecksumFileSerializer,
+        url_path=r'item/(?P<index>\d+)',
+    )
+    def item(self, request, pk, index):
+        collection = models.Collection.objects.get(pk=pk)
+        files = collection.checksumfiles.order_by('pk')
+        try:
+            instance = files[int(index)]
```

Comment:
@mcovalt, can you think of a better or different way to do this?

I'm wondering if `instance = files[int(index)]` is an expensive call - i.e., does it retrieve ALL ChecksumFiles and then grab the index? Or is there some django-magic happening where that's really just an iterator and `__getitem__` is perform a single query for that single object?

Reply:
Playing around in the shell:

```py
In [1]: collection = Collection.objects.first()
SELECT t.oid,
       typarray
  FROM pg_type t
  JOIN pg_namespace ns
    ON typnamespace = ns.oid
 WHERE typname = 'hstore'

Execution time: 0.001136s [Database: default]
SELECT typarray
  FROM pg_type
 WHERE typname = 'citext'

Execution time: 0.000602s [Database: default]
SELECT "rgd_collection"."id",
       "rgd_collection"."name",
       "rgd_collection"."description"
  FROM "rgd_collection"
 ORDER BY "rgd_collection"."id" ASC
 LIMIT 1

Execution time: 0.001140s [Database: default]

In [2]: ChecksumFile.objects.update(collection=collection)
UPDATE "rgd_checksumfile"
   SET "collection_id" = 1

Execution time: 0.008151s [Database: default]
Out[2]: 21

In [3]: files = collection.checksumfiles.order_by('pk')

In [5]: files[int(3)]
SELECT "rgd_checksumfile"."id",
       "rgd_checksumfile"."created",
       "rgd_checksumfile"."modified",
       "rgd_checksumfile"."failure_reason",
       "rgd_checksumfile"."status",
       "rgd_checksumfile"."name",
       "rgd_checksumfile"."description",
       "rgd_checksumfile"."checksum",
       "rgd_checksumfile"."validate_checksum",
       "rgd_checksumfile"."last_validation",
       "rgd_checksumfile"."collection_id",
       "rgd_checksumfile"."created_by_id",
       "rgd_checksumfile"."file_set_id",
       "rgd_checksumfile"."type",
       "rgd_checksumfile"."file",
       "rgd_checksumfile"."url"
  FROM "rgd_checksumfile"
 WHERE "rgd_checksumfile"."collection_id" = 1
 ORDER BY "rgd_checksumfile"."id" ASC
 LIMIT 1
OFFSET 3

Execution time: 0.002382s [Database: default]
Out[5]: <ChecksumFile: ChecksumFile object (4)>

In [6]:
```

It looks like the query has: ` LIMIT 1 OFFSET 3` so this should be a very inexpensive operation

Reply:
Nice! Yeah, Django's ORM is pretty great.